File: LeetCode/1_Two_Sum.go
// 1. 완전탐색 => Time: O(n^2), Space: O(1)
func twoSum(nums []int, target int) []int {
    for i, num := range nums {
        for j := i + 1; j < len(nums); j++ {
            if num + nums[j] == target {
                return []int{i, j}
            }
        }
    }
    return []int{}
}

// 2. Map 이용 => Time: O(n), Space: O(n)
// map에 key: num, value: index로 저장
// nums를 반복문을 돌면서 num을 하나씩 순회하는데, num과 더해서 target이 되려면 필요한 수는 target-num
// map에 target-num이 존재하면 사용할 수 있는 수이므로 num과 target-num의 index를 리턴한다.
// 존재하지 않으면 현재 순회한 num과 그 index를 map에 저장하여 나중에 검색할 수 있도록 한다.

// if문에 간단한 문장(optional statement)를 함깨 쓸 수 있다.
// map에 해당 key가 존재하는지는 map[key]의 두번째 리턴값을 가지고 확인할 수 있다.
// map[key]의 첫번째 리턴값: value, 두번째 리턴값: 해당 key가 존재하는지를 의미하는 boolean 값
func twoSum(nums []int, target int) []int {
    indices := make(map[int]int)

    for i, num := range nums {
        if idx, isPresent := indices[target-num]; isPresent {
            return []int{i, idx}
        }
        indices[num] = i
    }

    return []int{}
}

File: src/LeetCode/1_Two_Sum.cpp
// 1. 완전탐색 => Time: O(n^2), Space: O(1)
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> result;

        for (int i = 0; i < nums.size(); i++) {
            for (int j = i+1; j < nums.size(); j++) {
                if (nums[i] + nums[j] == target) {
                    result.push_back(i);
                    result.push_back(j);
                    return result;
                }
            }
        }
        return result;
    }
};


// 2. Map 이용 => Time: O(n), Space: O(n)
// map에 key: num, value: index로 저장
// nums를 반복문을 돌면서 num을 하나씩 순회하는데, num과 더해서 target이 되려면 필요한 수는 target-num
// map에 target-num이 존재하면 사용할 수 있는 수이므로 num과 target-num의 index를 리턴한다.
// 존재하지 않으면 현재 순회한 num과 그 index를 map에 저장하여 나중에 검색할 수 있도록 한다.

// 32-34 라인: 한 번에 묶어서 입출력
// {}로 vector를 바로 만들기 가능
// contains() 함수로 key가 map에 존재하는지 확인 가능
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        ios::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        unordered_map<int, int> index;

        for (int i = 0; i < nums.size(); i++) {
            int num = target - nums[i];

            if (index.contains(num)) {
                return {i, index[num]};
            }
            index[nums[i]] = i;
        }
        return {-1};
    }
};
